src/bin/ccp.rs

```diff
@@ -0,0 +1,407 @@
+#![recursion_limit = "128"]
+
+extern crate colored;
+extern crate libloading;
+extern crate portus;
+extern crate proc_macro2;
+extern crate quote;
+extern crate regex;
+extern crate structopt;
+extern crate syn;
+extern crate toml;
+extern crate walkdir;
+
+use colored::Colorize;
+use walkdir::WalkDir;
+//use clap::{App, Arg};
+use std::process::Command;
+
+#[derive(Debug)]
+struct Alg {
+    crate_name: String,
+    crate_path: String,
+}
+
+use toml::Value;
+
+fn find_algs(root: PathBuf) -> Vec<Alg> {
+    WalkDir::new(root)
+        .max_depth(1)
+        .min_depth(1)
+        .into_iter()
+        .filter_map(|entry| {
+            let entry = match entry {
+                Ok(ent) => ent,
+                Err(err) => {
+                    // TODO prettier
+                    println!("warn! unable to open entry: {:#?}", err);
+                    return None;
+                }
+            };
+
+            if !entry.file_type().is_dir() {
+                return None;
+            }
+
+            if entry.file_name() == "lib" {
+                return None;
+            }
+
+            //let crate_name = entry.path().file_name().unwrap().to_str().unwrap();
+            let crate_path = entry.path().to_str().unwrap();
+            let config_path = Path::new(entry.path()).join("Cargo.toml");
+            let config_str = std::fs::read_to_string(config_path.clone()).unwrap_or_else(|_| {
+                eprintln!(
+                    "==> error reading {}",
+                    config_path.to_string_lossy().red().bold()
+                );
+
+                std::process::exit(1);
+            });
+
+            let config_toml: Value = toml::from_str(&config_str).unwrap_or_else(|_| {
+                eprintln!(
+                    "==> error parsing {}",
+                    config_path.to_string_lossy().red().bold()
+                );
+
+                std::process::exit(1);
+            });
+
+            let crate_name = &config_toml["package"]["name"].as_str().unwrap();
+            Some(Alg {
+                crate_name: crate_name.to_string(),
+                crate_path: crate_path.to_owned(),
+            })
+        })
+        .collect()
+}
+
+fn generate_cargo_toml(algs: &Vec<Alg>) -> String {
+    let mut toml = format!(
+        r#"[package]
+name = "spawnccp"
+version = "0.1.0"
+edition = "2018"
+
+[lib]
+name = "spawnccp"
+crate-type = ["staticlib", "cdylib"]
+
+[dependencies]
+libc = "0.2"
+clap = "2.32"
+portus = "^0.4.3"
+"#
+    );
+
+    for Alg {
+        crate_name,
+        crate_path,
+    } in algs
```

Comment:
this formatting seems weird to me, but I guess it makes sense, and might as well follow rustfmt

Reply:
hm agreed, i didn't do this, rustfmt did :P i don't particularly care either way